src/orion/client/runner.py

```diff
@@ -0,0 +1,339 @@
+                except InvalidResult as exception:
+                    # stop the optimization process if we received `InvalidResult`
+                    # as all the trials are assumed to be returning those
+                    to_be_raised = exception
+                    self.client.release(trial, status="broken")
+
+            if isinstance(result, AsyncException):
+                if (
+                    isinstance(result.exception, ExecutionError)
+                    and result.exception.return_code == self.interrupt_signal_code
+                ):
+                    to_be_raised = KeyboardInterrupt()
+                    self.client.release(trial, status="interrupted")
+                    continue
+
+                # Regular exception, might be caused by the choosen hyperparameters
+                # themselves rather than the code in particular (like Out of Memory error
+                # for big batch sizes)
+                exception = result.exception
+                self.worker_broken_trials += 1
+                self.client.release(trial, status="broken")
+
+                if self.on_error is None or self.on_error(
+                    self, trial, exception, self.worker_broken_trials
+                ):
+                    log.error(result.traceback)
+
+                else:
+                    log.error(str(exception))
+                    log.debug(result.traceback)
+
+        # if we receive too many broken trials, it might indicate the user script
+        # is broken, stop the experiment and let the user investigate
+        if self.is_broken:
+            to_be_raised = BrokenExperiment(
+                "Worker has reached broken trials threshold"
+            )
+
+        if to_be_raised is not None:
+            log.debug("Runner was interrupted")
+            self._release_all()
+            raise to_be_raised
+
+        return len(results)
+
+    def _release_all(self):
+        """Release all the trials that were reserved by this runner.
+        This is only called during exception handling to avoid retaining trials
+        that cannot be retrieved anymore
+
+        """
+        # Sanity check
+        for _, trial in self.pending_trials.items():
+            try:
+                self.client.release(trial)
+            except AlreadyReleased:
+                pass
+
+        self.pending_trials = dict()
+
+    def _suggest_trials(self, count):
+        """Suggest a bunch of trials to be dispatched to the workers"""
+        trials = []
+        for _ in range(count):
+            try:
+                batch_size = count if self.batch_size == 0 else self.batch_size
+                trial = self.client.suggest(pool_size=batch_size)
+                trials.append(trial)
+
+            # non critical errors
+            except WaitingForTrials:
+                break
+
+            except ReservationRaceCondition:
+                break
+
+            except CompletedExperiment:
```

Comment:
Not tested

Reply:
Done here https://github.com/Epistimio/orion/pull/684/commits/0d390c553d734d53440bff566f85eb124b30152e